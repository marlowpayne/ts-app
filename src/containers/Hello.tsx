import React from 'react';
import { connect, Dispatch } from 'react-redux';
import styled from 'styled-components';

import { Hello as VisualComp } from '../visuals/Hello';
import { TextField } from '../visuals/TextField';
import { StoreState } from '../types';
import { defaultName, updateName, Action } from '../actions';

export interface Props {
  name: string;
  onSubmit: Function;
  onEmptySubmit: Function;
}

const Wrapper = styled.div``;

// exporting for testing purposes
export const Visual = (props: Props) => {
  const onSubmit = (text: string) => {
    if (!text) {
      props.onEmptySubmit();
    } else {
      props.onSubmit(text);
    }
  };

  return (
    <Wrapper>
      <VisualComp name={props.name} />
      <TextField {...{ onSubmit }} />
    </Wrapper>
  );
};

const mapStateToProps = ({ person }: StoreState) => ({
  name: person.get('name')
});

const mapDispatchToProps = (dispatch: Dispatch<Action>) =>
  ({
    onSubmit: (text: string) => dispatch(updateName(text)),
    onEmptySubmit: () => dispatch(defaultName()),
  });

export const Hello = connect(mapStateToProps, mapDispatchToProps)(Visual);
