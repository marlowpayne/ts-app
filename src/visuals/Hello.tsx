import React from 'react';
import styled from 'styled-components';

export interface Props {
  name?: string;
  enthusiasmLevel?: number;
}

const Wrapper = styled.div`
  text-align: center;
  margin: 20px;
  font-size: 48px;
  font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif;
`;

const Greeting = styled.div``;

export const Hello = ({ name = 'User', enthusiasmLevel = 1 }: Props) => {
  const getExclamationMarks = (numChars: number) => {
    return Array(numChars + 1).join('!');
  };

  if (enthusiasmLevel <= 0) {
    throw new Error('You could be a little more enthusiastic. :D');
  }

  return (
    <Wrapper>
      <Greeting className="greeting">
        Hello {name + getExclamationMarks(enthusiasmLevel)}
      </Greeting>
    </Wrapper>
  );
};
