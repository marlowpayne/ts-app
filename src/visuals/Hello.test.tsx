import React from 'react';
import { render } from 'enzyme';
import renderer from 'react-test-renderer';

import { Hello } from './Hello';

describe('Hello visual comp', () => {
  it('renders the correct text when no enthusiasm level is given', () => {
    const hello = render(<Hello name="Tester" />);
    expect(hello.find('.greeting').text()).toEqual('Hello Tester!');
  });

  it('renders the correct text with an explicit enthusiasm of 1', () => {
    const hello = render(<Hello name="Tester" enthusiasmLevel={1} />);
    expect(hello.find('.greeting').text()).toEqual('Hello Tester!');
  });

  it('renders the correct text with an explicit enthusiasm level of 5', () => {
    const hello = render(<Hello name="Tester" enthusiasmLevel={5} />);
    expect(hello.find('.greeting').text()).toEqual('Hello Tester!!!!!');
  });

  it('throws when the enthusiasm level is 0', () => {
    expect(() => {
      render(<Hello name="Tester" enthusiasmLevel={0} />);
    }).toThrow();
  });

  it('throws when the enthusiasm level is negative', () => {
    expect(() => {
      render(<Hello name="Tester" enthusiasmLevel={-1} />);
    }).toThrow();
  });
});

describe('Hello visual comp -- Snapshot', () => {
  it('captures a snapshot', () => {
    const rendered = renderer.create(<Hello name="Tester" enthusiasmLevel={5} />).toJSON();
    expect(rendered).toMatchSnapshot();
  });
});
