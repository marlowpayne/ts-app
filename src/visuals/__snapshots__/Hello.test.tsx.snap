// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hello visual comp -- Snapshot captures a snapshot 1`] = `
<div
  className="sc-bdVaJa jdlAAa"
>
  <div
    className="greeting sc-bwzfXH uTxCW"
  >
    Hello 
    Tester!!!!!
  </div>
</div>
`;
